### ISSUE TITLE
[DISCUSS] Remove notebook dep from jupyterlab and jupyterlab_server 

### ISSUE THREAD

**Issue by echarles** [jupyterlab/jupyterlab#7310]
[jupyter_server](https://github.com/jupyter/jupyter_server) is now synced with [notebook](https://github.com/jupyter/notebook).

This issue is to discuss options to move forward the jupyterlab server components (including [jupyterlab_server](https://github.com/jupyterlab/jupyterlab_server)).

I POC the complete notebook removal in 2 steps:
1. Use jupyter_server without the extensions mecanism.
2. Migrate to the jupyter_server extensions mecanism.

Any inputs on this initiative?

I am also wondering why we need to keep a separated [jupyterlab_server](https://github.com/jupyterlab/jupyterlab_server) - Could the functionality that repo bring could be intergrated in jupyterlab repo?

**Comment by jasongrout** [comment]
We have a separate jupyterlab_server package so that people can build custom fixed JupyterLab distributions without having a lot of the complexity that JupyterLab has with user-installable extensions. They can build on top of jupyterlab_server directly. For example, see our example app: https://github.com/jupyterlab/jupyterlab/tree/master/examples/app

**Comment by echarles** [comment]
@jasongrout Yes, I am fan of the `examples/app` and it makes perfectly sense to have such an easy way to build a jlab from scratch.

However, the `jupyterlab_server` repo contains some handlers (theme...) which a mandatory for jupyterlab. Shouldn't be moved to jupyterlab so they can be modified and releases with the other handlers? Going further, `the LabServerApp` could be a subpackage of jupyterlab.

Assuming we take the jupyter_server extension road, having one repo would make life easier for everyone. Of course, for now, this can perfectly remain separated, it is just that I bring this on the table as I have to touch those 2 repos now...

The main point for this discussion remains: should we migrate to `server` only and not rely anymore on `notebook`?

**Comment by jasongrout** [comment]
> The main point for this discussion remains: should we migrate to `server` only and not rely anymore on `notebook`?

1. I think `server` would need a stable release.
2. This would need to happen on a jlab major release (next is 2.0, slated for January)
3. A reason to have jlab depend on notebook is that is easy to switch back and forth between jlab and classic notebook. Do we have a timeline for the notebook to be converted to a server extension and rely on the `server` package?

**Comment by vidartf** [comment]
Note that a weekly call about jupyter server is set to start from *today*. See details here: https://github.com/jupyter/jupyter_server/issues/111

**Comment by echarles** [comment]
> I think server would need a stable release.

+1 - Working on that, see also today meeting linked by @vidartf 

> This would need to happen on a jlab major release (next is 2.0, slated for January)

Good to know. Also easier to deal with defined deadlines :)

> A reason to have jlab depend on notebook is that is easy to switch back and forth between jlab and classic notebook. Do we have a timeline for the notebook to be converted to a server extension and rely on the server package?

Yeah, the magic switch with the `/tree` url... One way would be `jlab -> notebook -> server`. Notebook migration to server is PRed in [notebook/4653](https://github.com/jupyter/notebook/pull/4653) and discussed today on zoom [jupyter_server#111](https://github.com/jupyter/jupyter_server/issues/111).

The outcome of that discussion could be go, no-go, wait and re-think/re-discuss.... For jupyterlab, it becomes a trade-off to decide if it is better to wait on a decision, or better to already migrate to jupyter-server, probably loosing the ability for end-user to switch back to the classic notebook...

**Comment by echarles** [comment]
Devil's advocate may state: `JLab 1.0 supports both Lab UI and classic notebook to allow smooth user migration - JLab 2.0 only offers the Lab UI and no more the classic notebook`

**Comment by echarles** [comment]
@jasongrout I had another look to the handlers in both repos.

- I see jupyterlab mainly contains code to build the source and manage the extensions.
- There is also in jupyterlab the `LabApp` which is run I believe when you invoke `jupyter lab` - There I am wondering how duplicate/different this `LabApp` is with the `LabServerApp` that resides in `jupyterlab_server`?

**Comment by echarles** [comment]
I have experimented a bit more 

**1. Experiment 1 - jupyterlab_server as jupyter_server extension invoked via the examples app**

Has been demoed yesterday at weekly server meeting organised by @Zsailer (just after @vidartf had to leave). Entrypoints were missing (as examples are invoked via python main.py) and extension/build handlers were also missing (the jupyterlab_server/app.py only provides the core handlers, not the extension handler which resided in jupyterlab repo, hence the second experiment)

[jupyterlab/server-ext](https://github.com/datalayer-contrib/jupyterlab/tree/server-ext)
[jupyterlab_server/server-ext](https://github.com/datalayer-contrib/jupyterlab-server/tree/server-ext)

The goal is to be able to run the 2 server extensions (notebook + lab) at the same time to offer to the endusers as now both classic notebook and jupyterlab - See example of 2 simultaneous server extensions in [simple example extension](https://github.com/jupyter/jupyter_server/pull/117).

**2. Experiment 2 - Move the Handlers from jupyterlab to jupyterlab_server**

Why? to have the experiment 1 being able to support the full jupyterlab runtime as a server.

[jupyterlab/handlers2server](https://github.com/datalayer-contrib/jupyterlab/tree/handlers2server)
[jupyterlab_server/handlers2servert](https://github.com/datalayer-contrib/jupyterlab-server/tree/handlers2server)

This experiment 2 does not implement the jupyter_server extension mechanism. Before going that way, I take a break to get feedback from the community.

One question: Would it make sense to merge back jupyterlab_server in jupyterlab. I can understand it is nice to have a light package without js artifacts to be used by a embeddeder (like the example app) - Would a tuned build be able to publish 2 artifacts (with and without js) from the same repo?

**Comment by echarles** [comment]
**3. Experiment 3 - Merge jupyterlab_server into jupyterlab and convert LabApp as a jupyter_server extension**

[jupyterlab/full-server-extension](https://github.com/datalayer-contrib/jupyterlab/tree/full-server-extension).

This branch is the jupyterlab repo with the jupyterlab_server python code and the `LabApp` converted to a jupyter_server extension, and notebook deps removed.

You can run the lab with `jupyter lab` or `jupyter server --ServerApp.jpserver_extensions="{'jupyterlab': True}"`

**Comment by echarles** [comment]
**4. Experiment 4 - CoRun Classic Notebook and JupyterLab as 2 jupyter_server extensions**

Launch the [jupyterlab/full-server-extension](https://github.com/datalayer-contrib/jupyterlab/tree/full-server-extension) with [jupyterlab/full-server-extension](https://github.com/datalayer-contrib/jupyterlab/tree/full-server-extension) and [/notebook//notebook-ext](https://github.com/Zsailer/notebook/tree/notebook-ext).

```bash
jupyter server --ServerApp.jpserver_extensions="{'notebook': True, 'jupyterlab': True}"
open http://localhost:8888/lab # Full blown JLab is working
open http://localhost:8888/lab # Full blown Classic Notebook is working
```

Thx @Zsailer for the hint on multiple server extensions.

This shows that we could further process migrating JLab and Classic Notebook to jupyter server extensions without loosing the ability to allow the user to switch back from jlab to classic.

The nice think is also that the launcher can even dissallow the switch by setting `--ServerApp.jpserver_extensions="{'notebook': False, 'jupyterlab': True}"`.

**Comment by echarles** [comment]
**5. Experiment 5 - JupyterLab as server Extension (without changing the current repos structure)**

Converging to something more acceptable (no change in current repos structure).

[jupyterlab/full-server-extension2](https://github.com/datalayer-contrib/jupyterlab/tree/full-server-extension2)

Will remove hardcoded paths and will look to migrate the other App (Build...) to extensions.

As this is a subject to discussion, I will join the next weekly meeting where hopefully we can put this topic at the agend (cc/ @jasongrout)

**Comment by blink1073** [comment]
Thanks for pushing on this @echarles!  The major concern we had raised in the first Jupyter Server meeting was in the handling of existing config files targeting the old package and app name structure.    The same shimming techniques could be used by both Notebook and JupyterLab.   
As far as combining the repos, I agree it would be a lot easier conceptually to have one repo.  
My concern there is that I've not seen good examples in the wild of releasing two Python packages from the same source tree.

**Comment by Zsailer** [comment]
Yes, @echarles, this is fantastic work! Thank you for working on this! 

I've created a new repo, [Zsailer/nbclassic](https://github.com/Zsailer/nbclassic/), that I'll be using as a playground for doing the notebook shims. It doesn't work right now, but I will ping this thread once it's ready for some testing. 

My goal is to create general shimming logic (for config and API changes) that JupyterLab may be able to copy.

@jasongrout may be able to comment on the idea of combining the two repos. I seem to remember he had a specific reason to keep the server logic in a separate place.

**Comment by jasongrout** [comment]
> @jasongrout may be able to comment on the idea of combining the two repos. I seem to remember he had a specific reason to keep the server logic in a separate place.

Mainly so that is easy to create custom opinionated JupyterLab distributions, like shown in the example app. These would be managed by the system admins, for example, with a defined set of supported plugins, with users not able to install their own plugins.

Turns out that the extension building mechanism adds a lot of complication to the python side of JupyterLab, so if you don't need it (or don't want it), it cuts down on the complexity quite a bit to build on the jupyterlab_server repo.

**Comment by echarles** [comment]
@blink1073 @jasongrout @Zsailer  Thx for your inputs.

I will leave the repos as it is for now and will create a PR with the removal of the notebook dep and migrating the Lab*App as jupyter_sever extension without chaning the repo structure. I have not a clear idea for the shim, so when @Zsailer will have implemented the shim in nbclassic, I will steal his work and inject it in that PR.

Sidenote on the repos structure:  For now the example app does not have the extensions nor build handlers, which make it a partial jupyterlab. 

- I have experimented migrating the handlers from the jupyterlab to jupyterlab_server. I had to workaround a few issues (e.g. the HERE and YARN_PATH variables), but it worked out. Moving only the handlers to jupyterlab_server leave the python build complexity in jupyterlab.
- One repo publishing multiple packages. A quick google search has not given me a solution. Less easy than it looks...

**Comment by echarles** [comment]
Thx to all for your feedbacks and inputs. The conclusion is that jupyterlab will switch to notebook with the nbclassic shim.

Follow-up of this is in following PRs:
[[WIP] JupyterLab as Server Extension](https://github.com/jupyterlab/jupyterlab/pull/7416)
[[WIP] JupyterLab Server as Server Extension](https://github.com/jupyterlab/jupyterlab_server/pull/79)
